src/lib.rs

```diff
@@ -194,13 +194,59 @@ pub enum InterpolationType {
+    /// which contains the samples for a single channel.
+    /// TODO! The `active_channels_mask` is optional. (active_channels_mask: Option<&[bool]>)
+    /// Any channel marked as inactive by a false value will be skipped during processing
+    /// and the corresponding output will also be empty.
+    /// If None is given, the length of each input is used to determine the active channels.
+    /// Then if an input channel has zero length, this channel will be considered as inactive. 
+    //fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>>;
+    fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>> {
+        let mut wave_out = self.allocate_output_buffer();
+        let channel_mask = make_mask_from_buffers(wave_in);
+        self.process_into_buffer(wave_in, &mut wave_out, &channel_mask)?;
+        Ok(wave_out)
+    }
+
+    /// Resample a chunk of audio to a pre-allocated output buffer.
+    /// Use this in real-time applications where the unpredictable time required to allocate memory
+    /// is causing issues. If this is not a problem, then use the more convenient `process` method.
     ///
     /// The input data is a slice, where each element of the slice is itself referenceable as a slice
     /// ([`AsRef<[T]>`](AsRef)) which contains the samples for a single channel. Since [`Vec<T>`] implements
     /// [`AsRef<[T]>`](AsRef), the input may simply be `&*Vec<Vec<T>>`. The output data is a vector, where each element
     /// of the vector is itself a vector which contains the samples for a single channel.
-    fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>>;
+    /// The output will be resized to fit all the output samples.
+    /// To avoid allocations, make sure that the output has sufficient capacity.
+    /// `active_channels_mask` is used to mark channels as active or inactive.
+    /// Any channel marked as inactive by a false value will be skipped during processing,
+    /// and the corresponding output will be left unchanged.
+    fn process_into_buffer<V: AsRef<[T]>>(
+        &mut self,
+        wave_in: &[V],
+        wave_out: &mut [Vec<T>],
+        active_channels_mask: &[bool],
+    ) -> ResampleResult<()>;
+
+    /// Convenience method for allocating an output buffer suitable for use with `process_into_buffer`.
+    fn allocate_output_buffer(&self) -> Vec<Vec<T>> {
+        let (channels, out_len) = self.get_max_output_size();
+        let mut wave_out = Vec::with_capacity(channels);
+        for _ in 0..channels {
+            wave_out.push(Vec::with_capacity(out_len));
+        }
+        wave_out
+    }
+
+    /// Get the max output size the resampler can give, as (channels, frames).
+    /// Note that when adjusting the ratio of an asynchronous resampler, the maximum size can change.
+    fn get_max_output_size(&self) -> (usize, usize);
```

Comment:
Considering this caveat, I'm questioning whether it's a good idea for this function to be in the generic `Resampler` trait. I think it's only appropriate for the resamplers with fixed output sizes.

Reply:
After adjusting the rate you should just call the function again to get the updated size. You can also skip that, the resampler will resize and allocate more space if needed.

Reply:
How about making a new trait for the resamplers with fixed output sizes and moving this function to that trait?

Reply:
I don't quite see the problem here. This function just provides the max size the resampler can currently output. You need this in order to preallocate. For the fixed output size resamples that's trivial, but for the varying output size ones it's quite tricky. I'll add something like "Note that adjusting the resampling ratio may change the max output size. Call this function after adjusting to get the updated size"

Reply:
Okay, I think documenting this limitation would be sufficient. How about this language for the documentation:

"For the Resamplers with variable output sizes ([FftFixedIn] and [SincFixedIn]), the max output size is only guaranteed to be sufficient to prevent allocation within [process_into_buffer] until the resampling ratio is changed."

Reply:
Perfect!